File: src/context_compactor.py
"""
context_compactor.py

Auto-compacts conversation history when approaching context window limits.
Summarizes older messages via the same LLM, preserving key context.
"""

import logging
from typing import Dict, List, Optional

from src.model_context import get_context_length, estimate_tokens
from src.llm_core import llm_call_async
from src.endpoint_resolver import resolve_endpoint
from core.models import ChatMessage

logger = logging.getLogger(__name__)

COMPACT_THRESHOLD = 0.85  # Trigger compaction at 85% of context window
SUMMARY_MAX_TOKENS = 1024
SMALL_CONTEXT_LIMIT = 8192  # Models with context <= this get aggressive trimming

# Cursor-style self-summarization prompt — produces structured, dense summaries
SELF_SUMMARY_SYSTEM_PROMPT = """You are summarizing a conversation to preserve context after compaction. Produce a structured summary that lets the conversation continue seamlessly.

Use this format:

## Conversation Summary
**Turns summarized:** {count}  |  **Compactions so far:** {n}

### User Goal
One sentence describing what the user is trying to accomplish.

### What Was Done
- Bullet points of completed actions, decisions made, and key outputs
- Include specific file paths, function names, variable names, URLs, and config values
- Note any errors encountered and how they were resolved

### Current State
What is the system/code/task state right now? What was the last thing discussed?

### Pending / Next Steps
- What remains to be done
- Any open questions or blockers

### Key Context
- Important constraints, preferences, or decisions that must not be forgotten
- Specific values: model names, ports, paths, credentials references, versions

Keep the summary under 1000 tokens. Be dense — every token should carry information. Do not include pleasantries or meta-commentary."""


def _sanitize_tool_messages(msgs: List[Dict]) -> List[Dict]:
    """Drop orphaned `tool` messages and dangling assistant `tool_calls`.

    OpenAI's API requires every `role:"tool"` message to immediately
    follow an assistant message that carries `tool_calls` (or another
    tool message in the same batch). Front-trimming the history can cut
    the assistant `tool_calls` parent while keeping its tool responses,
    which triggers: "messages with role 'tool' must be a response to a
    preceding message with 'tool_calls'". This pass repairs that:
      - drops `tool` messages with no valid preceding tool_calls
      - drops assistant `tool_calls` messages whose tool responses were
        all trimmed away (some providers reject unanswered tool_calls)
    """
    # Pass 1: drop orphan tool messages.
    cleaned: List[Dict] = []
    in_batch = False  # are we right after an assistant tool_calls (or mid-batch)?
    for m in msgs:
        role = m.get("role")
        if role == "tool":
            if in_batch:
                cleaned.append(m)
            # else: orphan — drop
            continue
        if role == "assistant" and m.get("tool_calls"):
            in_batch = True
        else:
            in_batch = False
        cleaned.append(m)

    # Pass 2: drop assistant tool_calls messages that have NO following
    # tool response (dangling) — walk backwards so we know what follows.
    out: List[Dict] = []
    for i, m in enumerate(cleaned):
        if m.get("role") == "assistant" and m.get("tool_calls"):
            nxt = cleaned[i + 1] if i + 1 < len(cleaned) else None
            if not (nxt and nxt.get("role") == "tool"):
                # Dangling tool_calls — keep the message but strip the
                # tool_calls so it's a plain assistant turn (preserves any
                # text content the model produced alongside the calls).
                m = {k: v for k, v in m.items() if k != "tool_calls"}
                if not (m.get("content") or "").strip():
                    continue  # nothing left worth keeping
        out.append(m)
    return out


def trim_for_context(messages: List[Dict], context_length: int, reserve_tokens: int = 512) -> List[Dict]:
    """Trim system messages to fit within context_length.

    For small-context models, progressively strips:
    1. RAG/memory system messages (keep preset system prompt)
    2. Older conversation turns
    Reserves space for the response.
    """
    budget = context_length - reserve_tokens
    used = estimate_tokens(messages)
    if used <= budget:
        return messages

    logger.info(f"Trimming messages: {used} tokens > {budget} budget (ctx={context_length})")

    # Separate system messages from conversation.
    # Messages marked _protected (e.g. active document) are never trimmed.
    system_msgs = []
    protected_msgs = []
    convo_msgs = []
    for msg in messages:
        if msg.get("_protected"):
            protected_msgs.append(msg)
        elif msg.get("role") == "system":
            system_msgs.append(msg)
        else:
            convo_msgs.append(msg)

    # Protected messages count toward budget but are never dropped
    protected_tokens = estimate_tokens(protected_msgs)
    budget -= protected_tokens

    # Priority: keep first system msg (preset prompt), drop others (memory, RAG, memo)
    essential_system = system_msgs[:1] if system_msgs else []
    extra_system = system_msgs[1:]

    # Try dropping extra system messages one by one (from the end)
    trimmed = essential_system + convo_msgs
    if estimate_tokens(trimmed) <= budget:
        # Dropping extras was enough — try adding back some
        result = list(essential_system)
        for msg in extra_system:
            candidate = result + [msg] + convo_msgs
            if estimate_tokens(candidate) <= budget:
                result.append(msg)
            else:
                break
        return _sanitize_tool_messages(result + protected_msgs + convo_msgs)

    # Still too big — truncate the first system message (but keep more than 500 chars)
    if essential_system:
        sys_text = essential_system[0].get("content", "")
        if len(sys_text) > 2000:
            essential_system[0] = {"role": "system", "content": sys_text[:2000] + "\n[System prompt truncated for context limits]"}
            trimmed = essential_system + convo_msgs
            if estimate_tokens(trimmed) <= budget:
                return _sanitize_tool_messages(essential_system + protected_msgs + convo_msgs)

    # Still too big — drop older conversation turns BUT protect the last 10.
    # Hermes-style: recent context matters more than old context.
    PROTECT_RECENT = 10
    if len(convo_msgs) > PROTECT_RECENT:
        old_msgs = convo_msgs[:-PROTECT_RECENT]
        recent_msgs = convo_msgs[-PROTECT_RECENT:]
        while old_msgs and estimate_tokens(essential_system + old_msgs + recent_msgs) > budget:
            old_msgs.pop(0)
        convo_msgs = old_msgs + recent_msgs
    else:
        # Not enough messages to split — just trim from front
        while convo_msgs and estimate_tokens(essential_system + convo_msgs) > budget:
            convo_msgs.pop(0)

    result = _sanitize_tool_messages(essential_system + protected_msgs + convo_msgs)
    logger.info(f"Trimmed to {estimate_tokens(result)} tokens ({len(result)} messages)")
    return result


async def maybe_compact(
    session,
    endpoint_url: str,
    model: str,
    messages: List[Dict],
    headers: Optional[Dict] = None,
) -> tuple:
    """Check context usage and compact if above threshold.

    Returns (messages, context_length, was_compacted).
    """
    context_length = get_context_length(endpoint_url, model)
    used = estimate_tokens(messages)
    pct = (used / context_length) * 100 if context_length else 0

    if pct < COMPACT_THRESHOLD * 100:
        return messages, context_length, False

    logger.info(
        f"Context at {pct:.1f}% ({used}/{context_length} tokens) — compacting"
    )

    # Split into system preface and conversation
    system_msgs = []
    convo_msgs = []
    for msg in messages:
        if msg.get("role") == "system":
            system_msgs.append(msg)
        else:
            convo_msgs.append(msg)

    if len(convo_msgs) < 4:
        return messages, context_length, False

    # Split conversation: summarize older half, keep recent half
    split_point = len(convo_msgs) // 2
    older = convo_msgs[:split_point]
    recent = convo_msgs[split_point:]

    # Build the text to summarize
    convo_text = "\n".join(
        f"{msg['role'].upper()}: {msg.get('content', '')[:2000]}"
        for msg in older
    )

    # Count prior compactions from existing summary messages
    compaction_count = sum(
        1 for m in system_msgs
        if "[Conversation summary" in m.get("content", "")
    )

    # Use utility model if configured, otherwise fall back to session model
    util_url, util_model, util_headers = resolve_endpoint("utility")
    compact_url = util_url or endpoint_url
    compact_model = util_model or model
    compact_headers = util_headers if util_url else headers

    prompt = SELF_SUMMARY_SYSTEM_PROMPT.replace(
        "{count}", str(len(older))
    ).replace(
        "{n}", str(compaction_count + 1)
    )
    summary_messages = [
        {"role": "system", "content": prompt},
        {"role": "user", "content": convo_text},
    ]

    try:
        summary = await llm_call_async(
            compact_url,
            compact_model,
            summary_messages,
            temperature=0.2,
            max_tokens=SUMMARY_MAX_TOKENS,
            headers=compact_headers,
            timeout=30,
        )
    except Exception as e:
        logger.error(f"Compaction summary failed: {e}")
        return system_msgs + recent, context_length, False

    summary_msg = {
        "role": "system",
        "content": f"[Conversation summary — earlier messages were compacted]\n{summary}",
    }

    compacted = system_msgs + [summary_msg] + recent

    # Update session history to match
    _update_session_history(session, split_point, summary)

    new_used = estimate_tokens(compacted)
    logger.info(
        f"Compacted: {used} -> {new_used} tokens "
        f"({len(older)} messages summarized, {len(recent)} kept)"
    )

    return compacted, context_length, True


def _update_session_history(session, split_point: int, summary: str):
    """Update the in-memory session history after compaction."""
    if not session or not hasattr(session, "history"):
        return

    if split_point >= len(session.history):
        return

    # Keep the recent messages, prepend summary
    recent_history = session.history[split_point:]
    summary_msg = ChatMessage(
        role="system",
        content=f"[Conversation summary]\n{summary}",
        metadata={"compacted": True, "summarized_count": split_point},
    )
    new_history = [summary_msg] + recent_history
    try:
        from core.models import get_session_manager_instance
        manager = get_session_manager_instance()
    except Exception:
        manager = None
    if manager and getattr(session, "id", None):
        if manager.replace_messages(session.id, new_history):
            return
    session.history = new_history

File: rust/src/src/context_compactor.rs
// src/context_compactor.rs  <- src/context_compactor.py
//! context_compactor.py
//!
//! Auto-compacts conversation history when approaching context window limits.
//! Summarizes older messages via the same LLM, preserving key context.
//!
//! SCOPE: the full pipeline is translated here — `_sanitize_tool_messages` and
//! `trim_for_context` (pure logic + the module constants and self-summary
//! prompt), PLUS the async/LLM-bound `maybe_compact(...)` (summarizes the older
//! half of the conversation via `llm_call_async` + `resolve_endpoint("utility")`)
//! and the session-manager-bound `_update_session_history(...)` (swaps the live
//! session's history for `[summary] + recent`, via the `SessionPersistence`
//! manager when present, else a local update). `get_context_length` is the pure
//! known-windows lookup, so `maybe_compact` adds no hot-path network call.

use crate::core::models::{session_manager, ChatMessage, Session};
use crate::src::endpoint_resolver::resolve_endpoint;
use crate::src::llm_core::llm_call_async;
use crate::src::model_context::{estimate_tokens, get_context_length};
use crate::pylog as logger;
use indexmap::IndexMap;
use serde_json::{json, Map, Value};

pub const COMPACT_THRESHOLD: f64 = 0.85; // Trigger compaction at 85% of context window
pub const SUMMARY_MAX_TOKENS: i64 = 1024;
pub const SMALL_CONTEXT_LIMIT: i64 = 8192; // Models with context <= this get aggressive trimming

/// Cursor-style self-summarization prompt — produces structured, dense summaries.
pub const SELF_SUMMARY_SYSTEM_PROMPT: &str = r#"You are summarizing a conversation to preserve context after compaction. Produce a structured summary that lets the conversation continue seamlessly.

Use this format:

## Conversation Summary
**Turns summarized:** {count}  |  **Compactions so far:** {n}

### User Goal
One sentence describing what the user is trying to accomplish.

### What Was Done
- Bullet points of completed actions, decisions made, and key outputs
- Include specific file paths, function names, variable names, URLs, and config values
- Note any errors encountered and how they were resolved

### Current State
What is the system/code/task state right now? What was the last thing discussed?

### Pending / Next Steps
- What remains to be done
- Any open questions or blockers

### Key Context
- Important constraints, preferences, or decisions that must not be forgotten
- Specific values: model names, ports, paths, credentials references, versions

Keep the summary under 1000 tokens. Be dense — every token should carry information. Do not include pleasantries or meta-commentary."#;

/// `m.get(key)` Python truthiness: present AND non-empty/non-zero/non-false.
fn truthy(m: &Value, key: &str) -> bool {
    match m.get(key) {
        None | Some(Value::Null) => false,
        Some(Value::Bool(b)) => *b,
        Some(Value::Number(n)) => n.as_f64().map(|f| f != 0.0).unwrap_or(true),
        Some(Value::String(s)) => !s.is_empty(),
        Some(Value::Array(a)) => !a.is_empty(),
        Some(Value::Object(o)) => !o.is_empty(),
    }
}

/// `m.get("role")` as `&str` ("" when missing/non-string — only used for `==`).
fn role(m: &Value) -> &str {
    m.get("role").and_then(|v| v.as_str()).unwrap_or("")
}

/// `(m.get("content") or "")` as a string slice (non-string content -> "").
fn content_str(m: &Value) -> &str {
    m.get("content").and_then(|v| v.as_str()).unwrap_or("")
}

/// Drop orphaned `tool` messages and dangling assistant `tool_calls`.
///
/// OpenAI's API requires every `role:"tool"` message to immediately follow an
/// assistant message that carries `tool_calls` (or another tool message in the
/// same batch). Front-trimming the history can cut the assistant `tool_calls`
/// parent while keeping its tool responses; this pass repairs that.
pub fn _sanitize_tool_messages(msgs: &[Value]) -> Vec<Value> {
    // Pass 1: drop orphan tool messages.
    let mut cleaned: Vec<Value> = Vec::new();
    let mut in_batch = false; // right after an assistant tool_calls (or mid-batch)?
    for m in msgs {
        let r = role(m);
        if r == "tool" {
            if in_batch {
                cleaned.push(m.clone());
            }
            // else: orphan — drop
            continue;
        }
        in_batch = r == "assistant" && truthy(m, "tool_calls");
        cleaned.push(m.clone());
    }

    // Pass 2: drop assistant tool_calls messages that have NO following tool
    // response (dangling).
    let mut out: Vec<Value> = Vec::new();
    for i in 0..cleaned.len() {
        let mut m = cleaned[i].clone();
        if role(&m) == "assistant" && truthy(&m, "tool_calls") {
            let nxt = cleaned.get(i + 1);
            let followed_by_tool = nxt.map(|n| role(n) == "tool").unwrap_or(false);
            if !followed_by_tool {
                // Strip tool_calls so it's a plain assistant turn (preserving any
                // text content the model produced alongside the calls).
                if let Some(obj) = m.as_object_mut() {
                    obj.remove("tool_calls");
                }
                // if not (m.get("content") or "").strip(): continue
                if content_str(&m).trim().is_empty() {
                    continue; // nothing left worth keeping
                }
            }
        }
        out.push(m);
    }
    out
}

/// `estimate_tokens` over an owned message vec (the Python passes lists around).
fn est(msgs: &[Value]) -> i64 {
    estimate_tokens(msgs)
}

/// Trim system messages to fit within `context_length`.
///
/// For small-context models, progressively strips RAG/memory system messages
/// (keeping the preset system prompt) and older conversation turns. Reserves
/// space for the response.
pub fn trim_for_context(messages: &[Value], context_length: i64, reserve_tokens: i64) -> Vec<Value> {
    let mut budget = context_length - reserve_tokens;
    let used = est(messages);
    if used <= budget {
        return messages.to_vec();
    }

    logger::info(&format!(
        "Trimming messages: {used} tokens > {budget} budget (ctx={context_length})"
    ));

    // Separate system / protected / conversation. `_protected` messages count
    // toward budget but are never dropped.
    let mut system_msgs: Vec<Value> = Vec::new();
    let mut protected_msgs: Vec<Value> = Vec::new();
    let mut convo_msgs: Vec<Value> = Vec::new();
    for msg in messages {
        if truthy(msg, "_protected") {
            protected_msgs.push(msg.clone());
        } else if role(msg) == "system" {
            system_msgs.push(msg.clone());
        } else {
            convo_msgs.push(msg.clone());
        }
    }

    let protected_tokens = est(&protected_msgs);
    budget -= protected_tokens;

    // Keep first system msg (preset prompt), drop others (memory, RAG, memo).
    let mut essential_system: Vec<Value> = system_msgs.iter().take(1).cloned().collect();
    let extra_system: Vec<Value> = system_msgs.iter().skip(1).cloned().collect();

    // Try dropping extra system messages, then add back what fits.
    let trimmed = concat(&[&essential_system, &convo_msgs]);
    if est(&trimmed) <= budget {
        let mut result = essential_system.clone();
        for msg in &extra_system {
            let candidate = concat(&[&result, std::slice::from_ref(msg), &convo_msgs]);
            if est(&candidate) <= budget {
                result.push(msg.clone());
            } else {
                break;
            }
        }
        return _sanitize_tool_messages(&concat(&[&result, &protected_msgs, &convo_msgs]));
    }

    // Still too big — truncate the first system message (keep > 500 chars).
    if !essential_system.is_empty() {
        let sys_text = content_str(&essential_system[0]).to_string();
        if sys_text.chars().count() > 2000 {
            let head: String = sys_text.chars().take(2000).collect();
            essential_system[0] = json!({
                "role": "system",
                "content": format!("{head}\n[System prompt truncated for context limits]"),
            });
            let trimmed = concat(&[&essential_system, &convo_msgs]);
            if est(&trimmed) <= budget {
                return _sanitize_tool_messages(&concat(&[
                    &essential_system,
                    &protected_msgs,
                    &convo_msgs,
                ]));
            }
        }
    }

    // Still too big — drop older turns BUT protect the last 10.
    const PROTECT_RECENT: usize = 10;
    if convo_msgs.len() > PROTECT_RECENT {
        let split = convo_msgs.len() - PROTECT_RECENT;
        let mut old_msgs: Vec<Value> = convo_msgs[..split].to_vec();
        let recent_msgs: Vec<Value> = convo_msgs[split..].to_vec();
        while !old_msgs.is_empty()
            && est(&concat(&[&essential_system, &old_msgs, &recent_msgs])) > budget
        {
            old_msgs.remove(0); // pop(0)
        }
        convo_msgs = concat(&[&old_msgs, &recent_msgs]);
    } else {
        while !convo_msgs.is_empty() && est(&concat(&[&essential_system, &convo_msgs])) > budget {
            convo_msgs.remove(0); // pop(0)
        }
    }

    let result = _sanitize_tool_messages(&concat(&[&essential_system, &protected_msgs, &convo_msgs]));
    logger::info(&format!(
        "Trimmed to {} tokens ({} messages)",
        est(&result),
        result.len()
    ));
    result
}

/// `trim_for_context(messages, context_length)` with the Python default
/// `reserve_tokens=512`.
pub fn trim_for_context_default(messages: &[Value], context_length: i64) -> Vec<Value> {
    trim_for_context(messages, context_length, 512)
}

/// Concatenate message slices (the Python `a + b + c` list joins).
fn concat(parts: &[&[Value]]) -> Vec<Value> {
    let mut out = Vec::new();
    for p in parts {
        out.extend(p.iter().cloned());
    }
    out
}

/// `maybe_compact(session, endpoint_url, model, messages, headers)` — check
/// context usage and compact (LLM-summarize the older half) if above
/// `COMPACT_THRESHOLD`. Returns `(messages, context_length, was_compacted)`.
///
/// Faithful port of `src/context_compactor.py:175-273`. Below threshold (the
/// common case) returns the messages UNCHANGED, so wiring this into the chat
/// path never regresses normal conversations — only very long ones (relative to
/// the model's known context window) get summarized, matching Python.
pub async fn maybe_compact(
    session: &mut Session,
    endpoint_url: &str,
    model: &str,
    messages: &[Value],
    headers: &IndexMap<String, String>,
) -> (Vec<Value>, i64, bool) {
    let context_length = get_context_length(endpoint_url, model);
    let used = estimate_tokens(messages);
    // pct = (used / context_length) * 100 if context_length else 0
    let pct = if context_length != 0 {
        (used as f64 / context_length as f64) * 100.0
    } else {
        0.0
    };

    if pct < COMPACT_THRESHOLD * 100.0 {
        return (messages.to_vec(), context_length, false);
    }

    logger::info(&format!(
        "Context at {pct:.1}% ({used}/{context_length} tokens) — compacting"
    ));

    // Split into system preface and conversation (order-preserving).
    let system_msgs: Vec<Value> = messages.iter().filter(|m| role(m) == "system").cloned().collect();
    let convo_msgs: Vec<Value> = messages.iter().filter(|m| role(m) != "system").cloned().collect();

    if convo_msgs.len() < 4 {
        return (messages.to_vec(), context_length, false);
    }

    // Summarize the older half, keep the recent half.
    let split_point = convo_msgs.len() / 2;
    let older = &convo_msgs[..split_point];
    let recent = &convo_msgs[split_point..];

    // convo_text = "\n".join(f"{role.upper()}: {content[:2000]}" for m in older)
    let convo_text = older
        .iter()
        .map(|m| {
            let r = role(m).to_uppercase();
            let c: String = content_str(m).chars().take(2000).collect();
            format!("{r}: {c}")
        })
        .collect::<Vec<_>>()
        .join("\n");

    // Count prior compactions from existing summary system messages.
    let compaction_count = system_msgs
        .iter()
        .filter(|m| content_str(m).contains("[Conversation summary"))
        .count();

    // Use the utility model if configured, otherwise fall back to the session
    // model/endpoint. `resolve_endpoint("utility")` => (url?, model?, headers?).
    let (util_url, util_model, util_headers) = resolve_endpoint("utility", session.owner.as_deref());
    let util_url_truthy = util_url.as_deref().map(|u| !u.is_empty()).unwrap_or(false);
    // compact_url = util_url or endpoint_url ; compact_model = util_model or model
    let compact_url = match &util_url {
        Some(u) if !u.is_empty() => u.clone(),
        _ => endpoint_url.to_string(),
    };
    let compact_model = match &util_model {
        Some(m) if !m.is_empty() => m.clone(),
        _ => model.to_string(),
    };
    // compact_headers = util_headers if util_url else headers
    let compact_headers: IndexMap<String, String> = if util_url_truthy {
        util_headers.unwrap_or_default()
    } else {
        headers.clone()
    };

    // prompt = SELF_SUMMARY_SYSTEM_PROMPT.replace("{count}", len(older)).replace("{n}", count+1)
    let prompt = SELF_SUMMARY_SYSTEM_PROMPT
        .replace("{count}", &older.len().to_string())
        .replace("{n}", &(compaction_count + 1).to_string());
    let summary_messages = vec![
        json!({"role": "system", "content": prompt}),
        json!({"role": "user", "content": convo_text}),
    ];

    let summary = match llm_call_async(
        &compact_url,
        &compact_model,
        summary_messages,
        0.2,
        SUMMARY_MAX_TOKENS,
        compact_headers,
        30,
    )
    .await
    {
        Ok(s) => s,
        Err(e) => {
            // except Exception: log + return system_msgs + recent (drop older), no compaction flag.
            logger::error(&format!("Compaction summary failed: {e}"));
            return (concat(&[&system_msgs, recent]), context_length, false);
        }
    };

    let summary_msg = json!({
        "role": "system",
        "content": format!("[Conversation summary — earlier messages were compacted]\n{summary}"),
    });
    let compacted = concat(&[&system_msgs, &[summary_msg], recent]);

    // Persist the compaction to the live session history.
    _update_session_history(session, split_point, &summary);

    let new_used = estimate_tokens(&compacted);
    logger::info(&format!(
        "Compacted: {used} -> {new_used} tokens ({} messages summarized, {} kept)",
        older.len(),
        recent.len()
    ));

    (compacted, context_length, true)
}

/// `_update_session_history(session, split_point, summary)` — swap the in-memory
/// session history for `[summary] + recent`, persisting via the session manager
/// when available (else a local update). Faithful port of
/// `src/context_compactor.py:275-300`.
fn _update_session_history(session: &mut Session, split_point: usize, summary: &str) {
    // if split_point >= len(session.history): return
    if split_point >= session.history.len() {
        return;
    }
    // recent_history = session.history[split_point:]
    let recent_history: Vec<ChatMessage> = session.history[split_point..].to_vec();
    let mut metadata = Map::new();
    metadata.insert("compacted".to_string(), Value::Bool(true));
    metadata.insert(
        "summarized_count".to_string(),
        Value::Number((split_point as i64).into()),
    );
    let summary_msg = ChatMessage::new(
        "system",
        format!("[Conversation summary]\n{summary}"),
        Some(metadata),
    );
    let mut new_history = vec![summary_msg];
    new_history.extend(recent_history);

    // if manager and session.id: if manager.replace_messages(...): return
    if !session.id.is_empty() {
        if let Some(manager) = session_manager() {
            if manager.replace_messages(&session.id, new_history.clone()) {
                // The manager updated its stored copy; mirror it onto the live
                // `&mut Session` (in Python they are the same object).
                session.history = new_history;
                return;
            }
        }
    }
    // session.history = new_history
    session.history = new_history;
}

#[cfg(test)]
mod compaction_tests {
    use super::*;

    fn session_with_history(id: &str, history: Vec<ChatMessage>) -> Session {
        let n = history.len() as i64;
        Session {
            id: id.to_string(),
            name: "t".to_string(),
            endpoint_url: "http://127.0.0.1:1/v1/chat/completions".to_string(),
            model: "test-model".to_string(),
            rag: false,
            archived: false,
            headers: IndexMap::new(),
            history,
            owner: None,
            is_important: false,
            message_count: n,
        }
    }

    // The common case: a few short messages stay far below 85% of the context
    // window, so `maybe_compact` returns them UNCHANGED with `was_compacted=false`
    // and never reaches `resolve_endpoint` / `llm_call_async` (pure, no I/O).
    #[tokio::test]
    async fn maybe_compact_below_threshold_is_noop() {
        let mut sess = session_with_history("s1", vec![]);
        let url = "http://127.0.0.1:1/v1/chat/completions".to_string();
        let messages = vec![
            json!({"role": "system", "content": "be brief"}),
            json!({"role": "user", "content": "hi"}),
            json!({"role": "assistant", "content": "hello"}),
        ];
        let (out, ctx, was) =
            maybe_compact(&mut sess, &url, "test-model", &messages, &IndexMap::new()).await;
        assert!(!was, "below threshold => not compacted");
        assert_eq!(out, messages, "messages pass through unchanged");
        assert!(ctx > 0);
    }

    // `_update_session_history` swaps history for `[summary] + recent` with the
    // Python metadata (`compacted`, `summarized_count`). Empty id => the (absent)
    // manager is skipped and the local update is applied.
    #[test]
    fn update_session_history_swaps_to_summary_plus_recent() {
        let history = vec![
            ChatMessage::new("user", "m0", None),
            ChatMessage::new("assistant", "m1", None),
            ChatMessage::new("user", "m2", None),
            ChatMessage::new("assistant", "m3", None),
        ];
        let mut sess = session_with_history("", history);
        _update_session_history(&mut sess, 2, "THE SUMMARY");
        assert_eq!(sess.history.len(), 3); // [summary] + 2 recent
        assert_eq!(sess.history[0].role, "system");
        assert_eq!(sess.history[0].content, "[Conversation summary]\nTHE SUMMARY");
        let md = sess.history[0].metadata.as_ref().unwrap();
        assert_eq!(md.get("compacted"), Some(&Value::Bool(true)));
        assert_eq!(md.get("summarized_count"), Some(&Value::Number(2.into())));
        assert_eq!(sess.history[1].content, "m2");
        assert_eq!(sess.history[2].content, "m3");
    }

    #[test]
    fn update_session_history_noop_when_split_past_end() {
        let mut sess = session_with_history("", vec![ChatMessage::new("user", "only", None)]);
        _update_session_history(&mut sess, 5, "X");
        assert_eq!(sess.history.len(), 1);
        assert_eq!(sess.history[0].content, "only");
    }
}
